scripts/west_commands/runners/qemu.py

```diff
@@ -4,28 +4,210 @@
+from pathlib import Path
+import shlex
+import subprocess
+import sys
+import yaml
 
+from runners.core import ZephyrBinaryRunner, RunnerCaps, FileType
+
+DEFAULT_QEMU_EXE = 'qemu-system-arm_rt595.exe' if sys.platform == 'win32' else 'qemu-system-arm_rt595'
+DEFAULT_QEMU_GDB_PORT = 1234
+DEFAULT_QEMU_FIFO = None
 
 class QemuBinaryRunner(ZephyrBinaryRunner):
     '''Place-holder for QEMU runner customizations.'''
 
+    def __init__(self, cfg, machine, cpu, dtb, qemu_opts,
+                 commander=DEFAULT_QEMU_EXE,
+                 pipe=DEFAULT_QEMU_FIFO,
+                 gdbserver='',
+                 gdb_host='127.0.0.1',
+                 gdb_port=DEFAULT_QEMU_GDB_PORT,
+                 tui=False, tool_opt=[]):
+        super().__init__(cfg)
+        self.file = cfg.file
+        self.file_type = cfg.file_type
+        self.hex_name = cfg.hex_file
+        self.bin_name = cfg.bin_file
+        self.elf_name = cfg.elf_file
+        self.gdb_cmd = [cfg.gdb] if cfg.gdb else None
+        self.machine = machine
+        self.cpu = cpu
+        self.dtb = dtb
+        self.pipe = pipe
+        self.use_loader = False
+        self.qemu_opts_content = []
+        self.qemu_opts = qemu_opts
+        self.commander = commander
+        self.gdbserver = gdbserver
+        self.gdb_host = gdb_host
+        self.gdb_port = gdb_port
+        self.tui_arg = ['-tui'] if tui else []
+        self.tool_opt = []
+        for opts in [shlex.split(opt) for opt in tool_opt]:
+            self.tool_opt += opts
+
+
     @classmethod
     def name(cls):
         return 'qemu'
 
     @classmethod
     def capabilities(cls):
-        # This is a stub.
-        return RunnerCaps(commands=set())
+        return RunnerCaps(commands={'flash', 'debug', 'debugserver', 'attach'},
+                          tool_opt=True, file=True)
 
     @classmethod
     def do_add_parser(cls, parser):
-        pass                    # Nothing to do.
+        # Required:
+        parser.add_argument('--machine', required=True, help='machine name')
+
+        # Optional:
+        parser.add_argument('--cpu', required=False, dest='cpu',
+                            help='specifies a cpu type')
+        parser.add_argument('--tui', default=False, action='store_true',
+                            help='if given, GDB uses -tui')
+        parser.add_argument('--gdbserver', default='JLinkGDBServer',
+                            help='GDB server, default is JLinkGDBServer')
+        parser.add_argument('--gdb-host', default='',
+                            help='custom gdb host, defaults to the empty string '
+                            'and runs a gdb server')
+        parser.add_argument('--gdb-port', default=DEFAULT_QEMU_GDB_PORT,
+                            help='pyocd gdb port, defaults to {}'.format(
+                                DEFAULT_QEMU_GDB_PORT))
+        parser.add_argument('--dtb', required=False, dest='dtb',
+                            help='specifies a dtb file')
```

Comment:
What's going on here?

Reply:
in some qemu platfroms,  they required a dtb file to do the dynamic initailzaiton, so I put this as an option to qemu.

Reply:
Please handle this by supporting `--tool-opt` and using that as needed from board.cmake. We have some legacy exceptions, but in general the runners do not add individual options that just wrap options in the underlying tool. This starts a process with no end. Instead, the future-proof solution is to support `--tool-opt`.

Reply:
@mbolivar-ampere  the `--tool-opt` is reserved for command line, if I use `--tool-opt` then it will be overried by command line, it is not what I am expected.